### DISCUSSION TITLE
What is the performance gain after optimizations?

### DISCUSSION THREAD

**Discussion by sourabhyadav** [zhiqwang/yolov5-rt-stack#156] [Q&A]
## ❓ Questions and Help
What is the performance gain after optimization with this method?  
Is there any breakup based on different yolov5 models like s,m,l,x and based on different batch sizes?  

## Additional context  
It would be helpful to get a rough idea of performance gains on a given GPU. This can also be useful to choose different optimization frameworks like libtorch, torchserve, tensorRT etc.  

Thanks for this awesome repo.

**Comment by zhiqwang** [comment]
Hi @sourabhyadav  ,

This question is a bit complicated to answer. In general, it depends on your needs.

If you focus on model inference speed, tools for hardware-side acceleration already exist. For example, if you are using a NVIDIA GPU device, there is no doubt that you should choose [TensorRT](https://github.com/nvidia/tensorrt). If you are using an Intel CPU device, you should choose [OpenVINO](https://github.com/openvinotoolkit/openvino) as the first option.

On the other hand, [LibTorch/TorchScript](https://pytorch.org/tutorials/beginner/Intro_to_TorchScript_tutorial.html) has the highest degree of freedom. Most models on PyTorch can be easily converted into TorchScript (Modify with minimal c++ code). For example, if we want to train a model with high accuracy, one of the tricks is to maintain the scale invariance of the image, we call this as a **dynamic shape** inference. For existing inference backends within my knowledge, LibTorch is the best one to address this demand. Although TensorRT and ONNXRuntime have similar mechanism to deal with this scenario, they have some limitations. If you can tolerate the loss of time, LibTorch is a good option.

To sum up, the speed of program development and the efficiency of model inference need to be balanced.

**Comment by zhiqwang** [comment]
> Is there any breakup based on different yolov5 models like s, m, l, x and based on different batch sizes?

Yeah, the model architectures will affect the inference time. `yolov5s` is the fastest, but its accuracy is relatively the worst. On the other hand, `yolov5x` is the slowest with the most accurate.

The batch size will influence the RAM usage, the bigger batch-size, the higher the memory consumption, and relatively the faster inference time (per image).

**Comment by zhiqwang** [comment]
And to answer this question, it is necessary to explain my original intention of creating this repo and next step plans.

My main idea is to make the training and the inference better connect together.

We've done the LibTorch backend inference, and it can retain all the features of PyTorch.

I have some hesitation on the [ONNXRuntime](https://github.com/microsoft/onnxruntime), TensorRT and the OpenVINO backends. I've done some work paving the way for these backends. Actually, you could find various implementations of these on GitHub. Also because of my limited experience in C++, I'm not planning to develop in this direction recently. We are welcome and eager for contributions here.

Recently I am focusing on the training side with the help of [lightning](http://github.com/pytorchlightning/pytorch-lightning/), I wanna train a leading model from scratch as ultralytics/yolov5.

**Comment by lucasjinreal** [comment]
@zhiqwang How's the training scrath goes? I revisited your project and found your code style are more modularied to the ultralytics version. I wonder: 1) Do u reproduced AP as original yolov5 (by training); 2) Do u planned to complete the deploy stack as a strength of this repo by uniform deployment via ONNX to different platform?

**Comment by zhiqwang** [comment]
Hi @jinfagang 

> 1) Do u reproduced AP as original yolov5 (by training);

Currently, we have not reproduced the accuracy of ultralytics. It is quite difficult to reproduce a leading model, we are eager for partners to work together.

As I mentioned above, I am going to try to implement the trainer in the route of Lightning Callback. And I'm doing some paving work about the datasets pipeline now.

> 2) Do u planned to complete the deploy stack as a strength of this repo by uniform deployment via ONNX to different platform?

My current implementation here is mainly based on the upstream [torchvision](https://github.com/pytorch/vision/). Acutally if you can deploy a model of torchvision successfully, you can also deploy yolort to your device. In other words, we are following the way suggested by torchvision.

Thanks for the rapid progress of DL accelerators, the pain points for deploying a object detection task is actually the pre-processing and post-processing procedures now. In the details, different frameworks have different implementations. Let's take a closer look at the `nms` which is in the post-processing part. Within my limited knowledge, most frameworks such as NCNN and MNN will put `nms` outside the DL accelerator, but there are also some frameworks that put `nms` in the DL accelerator like LibTorch, ONNXRuntime and OpenVINO. (We have done the LibTorch and ONNXRuntime parts in this repo.)

It will be awesome if we could achieve a uniform deployment! Currently the inference frameworks are also evolving rapidly, achieving such a benchmark is full of challenges (Whether it’s through ONNX or not). For myself, the LibTorch and torchscript is convenient enough to solve most scenarios.

To sum up, I will focus on the first part in the next few months. I will also track the progress of the deployment, after all, it will attract more attention and be more practical.
